app/controllers/scim_rails/scim_groups_controller.rb

```diff
@@ -21,29 +21,233 @@ def index
       json_scim_group_response(object: groups, counts: counts)
     end
 
-    # TODO: complete method
     def create
-      group = @company.public_send(ScimRails.config.scim_group_scope).create!(permitted_group_params)
+      group_attributes = permitted_group_params(params)
 
-      json_scim_response(object: group, status: :created)
+      if ScimRails.config.scim_group_prevent_update_on_create
+        group = @company.public_send(ScimRails.config.scim_groups_scope).create!(group_attributes.except(:members))
+      else
+        username_key = ScimRails.config.queryable_group_attributes[:userName]
+        find_by_username = {}
+        find_by_username[username_key] = group_attributes[username_key]
+        group = @company
+          .public_send(ScimRails.config.scim_groups_scope)
+          .find_or_create_by(find_by_username)
+        group.update!(group_attributes.except(:members))
+      end
+
+      update_group_status(group) unless put_active_param.nil?
+      json_scim_group_response(object: group, status: :created)
     end
 
     def show
-      group = @company.public_send(ScimRails.config.scim_group_scope).find(params[:id])
-      json_scim_response(object: group)
+      group = @company.public_send(ScimRails.config.scim_groups_scope).find(params[:id])
+      json_scim_group_response(object: group)
     end
 
-    # TODO: complete method
     def put_update
+      group = @company.public_send(ScimRails.config.scim_groups_scope).find(params[:id])
+
+      put_error_check
+
+      group_attributes = permitted_group_params(params)
+      group.update!(group_attributes.except(:members))
 
+      update_group_status(group) unless put_active_param.nil?
+
+      member_ids = params["members"]&.map{ |member| member["value"] }
+
+      group.users.clear
+      add_members(group, member_ids)
+
+      json_scim_group_response(object: group)
     end
 
-    # TODO: complete method
     def patch_update
+      group = @company.public_send(ScimRails.config.scim_groups_scope).find(params[:id])
+
+      params["Operations"].each do |operation|
+        case operation["op"]
+        when "replace"
+          patch_replace(group, operation)
+        when "add"
+          patch_add(group, operation)
+        when "remove"
+          patch_remove(group, operation)
+        else
+          raise ScimRails::ExceptionHandler::UnsupportedGroupPatchRequest
+        end
+      end
 
+      json_scim_group_response(object: group)
     end
 
     private
 
+    def member_error_check(members)
+      raise ScimRails::ExceptionHandler::InvalidMembers unless (members.is_a?(Array) && array_of_hashes?(members))
+
+      member_ids = members.map{ |member| member["value"] }
+
+      member_ids.each do |id|
+        @company.public_send(ScimRails.config.scim_users_scope).find(id)
```

Comment:
What is the purpose of this?
It's doing a DB lookup for each member ID, but it just throws it away after

Reply:
It's essentially an error check that the user ids which are passed in are valid.  Without this check, the users in a group may be blown away for no reason.  

For example in the PUT method, it first clears the users in a group, and then adds members back to the group based on the ids which were passed into the method.  However if one of these ids are not valid, an error will be raised and the users will have been cleared from the group for no reason.  That was my reasoning in putting a check at the start of the method, so that situations such as this wouldn't happen.